crates/datafusion_ext/src/planner/relation/mod.rs

```diff
@@ -261,10 +268,45 @@ fn infer_func_for_file(path: &str) -> Result<OwnedTableReference> {
             schema: "public".into(),
             table: "read_bson".into(),
         },
+        "gz" => {
+            //handling compressed files with .gz extension
+            infer_func_from_compressed_file(filename)?
+        }
         ext => {
             return Err(DataFusionError::Plan(format!(
                 "unable to infer how to handle file extension: {ext}"
             )))
         }
     })
 }
+
+fn infer_func_from_compressed_file(filename: &str) -> Result<OwnedTableReference> {
+    if filename.contains(".json.gz")
+        | filename.contains(".json1.gz")
+        | filename.contains(".ndjson.gz")
+    {
+        return Ok(OwnedTableReference::Partial {
+            schema: "public".into(),
+            table: "ndjson_scan".into(),
+        });
+    } else if filename.contains(".parquet.gz") {
+        return Ok(OwnedTableReference::Partial {
+            schema: "public".into(),
+            table: "parquet_scan".into(),
+        });
+    } else if filename.contains(".csv.gz") {
+        return Ok(OwnedTableReference::Partial {
+            schema: "public".into(),
+            table: "csv_scan".into(),
+        });
+    } else if filename.contains(".bson.gz") {
+        return Ok(OwnedTableReference::Partial {
+            schema: "public".into(),
+            table: "read_bson".into(),
+        });
```

Comment:
I'm not 100% sure that this will actually work.

Reply:
I think we should still pass this in to the planner even if it's invalid. I don't think this stage of the planning should be concerned with plan validity. It's only concerned with creating the logical plan from a valid query, even if that plan itself is invalid.

Reply:
I'm not sure that makes a lot of sense, or at least that's not consistent with the implementation. If we want to implement what you describe than we split the right side of the string on `.` and match the penultimate phrase against the formats, and the ultimate segment on the compression formats. Really, If it's not _this_ code's responsibility to decide if the reading function can handle the compression formats, then we should _actually_ ignore them here.

But I don't think that's what we actually want, and I don't think "pass invalid arguments through and hope that the downstream code handles it correctly," is a reasonable design and also (more importantly) it makes it harder to return a clear error message to users in invalid  cases.

Reply:
It's not "hoping downstream code handles it correctly". It's a matter of separation of logic and keeping things DRY. The table function already need to handle this. There's no reason to check it here also. Otherwise we have 2 places to check identical logic. 

```sql
select * from read_bson('some_file.bson.gz');
select * from 'some_file.bson.gz';
```

Reply:
> It's a matter of separation of logic and keeping things DRY.

My concern is primarily for centralizing the encoding of the validation logic for all of the formats and table functions (and DDL-created dispatches to these table providers), so that we can have clear error messages, and avoid having duplicate validation logic.

> The table function already need to handle this scenario.

Sort of? The table functions are going to error if they can't read the data or if it's in the wrong format. It's maybe _not_ necessary for them to validate that the path itself is well-formed.

> It's redundant to check it here also.

It is redundant, but so is enumerating every possible compression algorithm and extension variant. We shouldn't need to care about `.bson.gz` vs `.bson.gzip` (and so forth). I think the reasonable solution to getting something here that's helpful and useful is:

- split the string by `.` if the last or second-to-last element in the resulting sequence is `bson`, `ndjson`, `jsonl`, or `csv` then dispatch to the appropriate function. 
- let's leave `.json` out of this as much as possible because of ambiguity between text sequences and line-separated json. 
- write tests to validate that the table functions behave correctly with compression and propagate reasonable error messages otherwise.